File: app/src/main/java/com/aol/philipphofer/logic/Timer.java
package com.aol.philipphofer.logic;

import java.util.Locale;

public class Timer {

    private final TimerListener timerListener;

    private boolean stopped = false;
    private boolean running = false;
    private int time; // in seconds

    public Timer(TimerListener timerListener) {
        this.timerListener = timerListener;
        this.init();
        setTime(0);
    }

    private void init() {
        new Thread() {
            @Override
            public void run() {
                while (!stopped) {
                    while (running) {
                        try {
                            sleep(1000);
                            if (running) {
                                setTime(++time);
                            }
                        } catch (InterruptedException e) {
                            e.printStackTrace();
                        }
                    }
                }
            }
        }.start();
    }

    private void setTime(int time) {
        this.time = time;
        this.timerListener.timeUpdate(this.time);
    }

    public int getTime() {
        return this.time;
    }

    public boolean isRunning() {
        return this.running;
    }

    public void startTimer(int time) {
        setTime(time);
        this.running = true;
    }

    public void startTimer() {
        this.running = true;
    }

    public void stopTimer() {
        this.running = false;
    }

    public void killTimer() {
        this.stopTimer();
        this.stopped = true;
    }

    public static String timeToString(int timeInSeconds) {
        int hours = timeInSeconds / 3600;
        int minutes = (timeInSeconds % 3600) / 60;
        int seconds = timeInSeconds % 60;

        if (hours > 0)
            return String.format(Locale.ENGLISH, "%d:%02d:%02d", hours, minutes, seconds);
        return String.format(Locale.ENGLISH, "%02d:%02d", minutes, seconds);
    }

    public interface TimerListener {
        void timeUpdate(int time);
    }
}

File: app/src/main/java/com/aol/philipphofer/logic/Timer.kt
package com.aol.philipphofer.logic

import java.util.Locale

class Timer(private val timerListener: TimerListener) {

    var isRunning = false
        private set
    var time = 0 // in seconds
        set(time) {
            field = time
            timerListener.timeUpdate(time)
        }
    private var stopped = false

    init {
        init()
        time = 0
    }

    private fun init() {
        object : Thread() {
            override fun run() {
                while (!stopped) {
                    while (isRunning) {
                        try {
                            sleep(1000)
                            if (isRunning) time = ++time
                        } catch (e: InterruptedException) {
                            e.printStackTrace()
                        }
                    }
                }
            }
        }.start()
    }

    fun startTimer(time: Int) {
        this.time = time
        isRunning = true
    }

    fun startTimer() {
        isRunning = true
    }

    fun stopTimer() {
        isRunning = false
    }

    fun killTimer() {
        stopTimer()
        stopped = true
    }

    interface TimerListener {
        fun timeUpdate(time: Int)
    }

    companion object {
        @JvmStatic
        fun timeToString(timeInSeconds: Int): String {
            val hours = timeInSeconds / 3600
            val minutes = timeInSeconds % 3600 / 60
            val seconds = timeInSeconds % 60
            return if (hours > 0) String.format(
                Locale.ENGLISH,
                "%d:%02d:%02d",
                hours,
                minutes,
                seconds
            ) else String.format(Locale.ENGLISH, "%02d:%02d", minutes, seconds)
        }
    }
}